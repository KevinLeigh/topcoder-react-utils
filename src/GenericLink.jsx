/**
 * The GenericLink wraps around React Router's Link or NavLink component, to
 * automatically replace them by regular <a> elements when:
 * - The target reference points to another domain;
 * - User opts to open the reference in a new tab;
 * - User explicitely opts to use <a>.
 */

/* global document, window */

import PT from 'prop-types';
import React from 'react';
import Url from 'url-parse';

export default function GenericLink(props) {
  const {
    activeClassName,
    activeStyle,
    children,
    className,
    enforceA,
    exact,
    isActive,
    location,
    onClick,
    onMouseDown,
    openNewTab,
    replace,
    routerLinkType,
    strict,
    to,
  } = props;
  /* Renders the link as <a> element if either opted explicitely, or the link
   * should open a new tab, or it is an anchor reference. */
  if (enforceA || openNewTab || to.startsWith('#')) {
    return (
      <a
        className={className}
        href={to}
        onClick={onClick}
        onMouseDown={onMouseDown}
        rel={openNewTab ? 'noopener noreferrer' : ''}
        target={openNewTab ? '_blank' : ''}
      >{children}</a>
    );
  }

  /* Otherwise we render the link as React Router's Link or NavLink element. */
  return React.createElement(
    routerLinkType, {
      activeClassName,
      activeStyle,
      className,
      exact,
      isActive,
      location,
      onClick: (e) => {
        /* If a custom onClick(..) handler was provided we execute it. */
        if (onClick) onClick(e);

        /* If the URL leads to a different domain we make transition using JS,
         * preventing event processing by React Router. */
        const url = new Url(to);
        if (url.origin !== document.location.origin) {
          document.location = to;
          e.preventDefault();

        /* Scroll to the top-left corner of the page, just in case the link
         * references the same page we are already in. */
        } else window.scroll(0, 0);
      },
      onMouseDown,
      replace,
      strict,
      to,
    },
  );
}

GenericLink.defaultProps = {
  activeClassName: null,
  activeStyle: null,
  children: null,
  className: null,
  enforceA: false,
  exact: false,
  isActive: null,
  location: null,
  onClick: null,
  onMouseDown: null,
  openNewTab: false,
  replace: false,
  strict: false,
  to: '',
};

GenericLink.propTypes = {
  activeClassName: PT.string,
  activeStyle: PT.string,
  children: PT.node,
  className: PT.string,
  enforceA: PT.bool,
  exact: PT.bool,
  isActive: PT.func,
  location: PT.shape(),
  onClick: PT.func,
  onMouseDown: PT.func,
  openNewTab: PT.bool,
  replace: PT.bool,
  routerLinkType: PT.element.isRequired,
  strict: PT.bool,
  to: PT.oneOfType([PT.object, PT.string]),
};
