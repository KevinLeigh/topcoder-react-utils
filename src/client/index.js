/**
 * Initialization of client-side code.
 */

/* global BUILD_INFO window */
window.TRU_BUILD_INFO = BUILD_INFO;
window.TRU_FRONT_END = true;
