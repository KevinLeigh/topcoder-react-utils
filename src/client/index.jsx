/**
 * Initialization of client-side code.
 */

/* global BUILD_INFO document window */

import forge from 'node-forge';
import React from 'react';
import ReactDom from 'react-dom';
import shortId from 'shortid';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';

/**
 * Renders given application into DOM, providing it with Redux store, if given.
 * @param {Function} Application Root ReactJS component of the app.
 * @param {Object} store Optional. Redux store.
 */
function render(Application, store) {
  let app = <BrowserRouter><Application /></BrowserRouter>;
  if (store) app = <Provider store={store}>{app}</Provider>;
  ReactDom.hydrate(app, document.getElementById('react-view'));
}

/**
 * Initializes the code at client side. It takes care about receiving the data
 * injected at the server-side, and also about setting up client side of hot
 * module reloading (HMR).
 * @param {Object} options Parameters accepted by the function.
 * @param {Function} options.Application Root component of the application to be
 *  rendered at the client side.
 * @param {String} options.appModulePath Optional. Path to the root module of
 *  the application. It should be provided to setup HMR.
 * @param {Function} options.customInit Optional. A custom function to be called
 *  before initial client-side rendering. It should return a promise that
 *  resolves to an object with the following params:
 *  - store {Object} - Optional. Redux store to be used.
 */
export default async function Init(options) {
  /* Injection of build-time data information into the client-side bundle. */
  window.TRU_BUILD_INFO = BUILD_INFO;
  window.TRU_FRONT_END = true;

  /* Removes data injection script out of the document. */
  const block = document.querySelector('script[id="inj"]');
  document.getElementsByTagName('body')[0].removeChild(block);

  /* Decodes data injected at the server side. */
  const { key } = window.TRU_BUILD_INFO;
  let data = forge.util.decode64(window.INJ);
  const decipher = forge.cipher.createDecipher('AES-CBC', key);
  decipher.start({ iv: data.slice(0, 32) });
  decipher.update(forge.util.createBuffer(data.slice(32)));
  decipher.finish();
  data = JSON.parse(forge.util.decodeUtf8(decipher.output.data));

  window.CONFIG = data.CONFIG;
  window.ISTATE = data.ISTATE;

  let store;
  if (options.customInit) ({ store } = await options.customInit());
  render(options.Application, store);

  if (module.hot && options.appModulePath) {
    module.hot.accept(
      options.appModulePath,
      () => render(options.Application, store),
    );

    /* HMR of CSS code each time webpack hot middleware updates the code. */
    /* eslint-disable no-underscore-dangle */
    const hotReporter = window.__webpack_hot_middleware_reporter__;
    /* eslint-enable no-underscore-dangle */
    const hotSuccess = hotReporter.success;
    hotReporter.success = () => {
      const stamp = shortId();
      const links = document.querySelectorAll('link[rel=stylesheet]');
      for (let i = 0; i < links.length; i += 1) {
        links[i].href = `${links[i].href.match(/[^?]*/)}?v=${stamp}`;
      }
      hotSuccess();
    };
  }
}
