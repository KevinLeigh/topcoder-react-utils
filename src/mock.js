const mock = require('./index');
mock.utils = require('./shared/__mocks__/utils').default;

module.exports = mock;
