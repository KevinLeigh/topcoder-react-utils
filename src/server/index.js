/**
 * Standard web servers.
 */

import atob from 'atob';
import _ from 'lodash';
import http from 'http';

import { isServerSide } from 'utils/isomorphy';

/* Polyfill required by ReactJS. */
import 'raf/polyfill';

import serverFactory from './server';

/* Polyfill for codes that rely on atob(..) function present in broswer
 * environments. */
global.atob = atob;

/* TODO: Should use the isClientSide(..) method from isomorphy module, once that
 * module is moved to topcoder-react-utils. */
if (!isServerSide()) {
  throw new Error('Cannot execute server-side code in other environment');
}

/**
 * Normalizes a port into a number, string, or false.
 * @param {String} value Port name or number.
 * @return Port number (Number), name (String), or false.
 */
function normalizePort(value) {
  const port = _.toNumber(value);
  if (_.isFinite(port)) return port; /* port number */
  if (!_.isNumber(port)) return value; /* named pipe */
  return false;
}

/**
 * Creates and starts a new webserver.
 * @param {Object} webpackConfig The Webpack config that was used to build the
 *  frontend bundle.
 * @param {config} options Additional options:
 *  - Application {Function} - Optional. The root ReactJS component of the app
 *    to use for server-side rendering;
 *  - devMode {Boolean} - Whether the server should be started in dev mode;
 *  - favicon {String} - Path of the favicon to be used by the server;
 *  - logger {Object} - Optional. The logger to use. By default, the console
 *    is used (which is not a good decision performancewise, but it will be
 *    changed soon);
 *  - beforeRender {Function} - The hook into server-side rendering. It will get
 *    incoming request as the argument and it should return a promise that will
 *    resolve to the object with the following fields all optional:
 *    - config {Object} - Config object to be injected into the page;
 *    - extraScripts {Array} - Any additional scripts to be injected into
 *      HTML template;
 *    - store {Object} - Redux store, which state should be used for server-side
 *      rendering, if it is performed, and also injected into HTML template as
 *      the initial state.
 *  - onExpressJsSetup {Function} - Custom setup of ExpressJS server.
 *  - port {String} - Optional. The port to listen (number or name). If not
 *    specified the value will be taken from PORT environmental variable, or
 *    default to 3000 otherwise.
 * @return {Promise} Resolves to the result object has two fields:
 *  - express {Object} - ExpressJS server;
 *  - http {Object} - NodeJS HTTP server.
 */
export default async function launch(webpackConfig, options) {
  /* Options normalization. */
  const ops = options ? _.clone(options) : {};
  ops.port = normalizePort(ops.port || process.env.PORT || 3000);
  _.defaults(ops, {
    logger: console,
  });

  /* Creates servers, resolves and sets the port. */
  const expressServer = await serverFactory(webpackConfig, ops);
  const httpServer = http.createServer(expressServer);

  /* Sets error handler for HTTP server. */
  httpServer.on('error', (error) => {
    if (error.syscall !== 'listen') throw error;
    const bind = _.isString(ops.port) ? `Pipe ${ops.port}` : `Port ${ops.port}`;

    /* Human-readable message for some specific listen errors. */
    switch (error.code) {
      case 'EACCES':
        ops.logger.error(`${bind} requires elevated privileges`);
        process.exit(1);
        break;
      case 'EADDRINUSE':
        ops.logger.error(`${bind} is already in use`);
        process.exit(1);
        break;
      default:
        throw error;
    }
  });

  /* Listening event handler for HTTP server. */
  httpServer.on('listening', () => {
    const addr = httpServer.address();
    const bind = _.isString(addr) ? `pipe ${addr}` : `port ${addr.port}`;
    ops.logger.info(
      `Server listening on ${bind} in ${process.env.NODE_ENV} mode`);
  });

  httpServer.listen(ops.port);

  return {
    expressServer,
    httpServer,
  };
}
