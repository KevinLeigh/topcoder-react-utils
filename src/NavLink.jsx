/**
 * The NavLink wraps around React Router's NavLink component, to automatically
 * replace it by the regular <a> element when:
 * - The target reference points to another domain;
 * - User opts to open the reference in a new tab;
 * - User explicitely opts to use <a>.
 */

import React from 'react';
import { NavLink as RrNavLink } from 'react-router-dom';

import GenericLink from './GenericLink';

export default function NavLink(props) {
  return <GenericLink {...props} routerLinkType={RrNavLink} />;
}

