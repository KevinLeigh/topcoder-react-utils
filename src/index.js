import client from 'client';
import Avatar from 'components/Avatar';
import Button from 'components/Button';
import Link from 'components/Link';
import NavLink from 'components/NavLink';
import ScalableRect from 'components/ScalableRect';
import * as utils from 'utils';

import 'styles/global.scss';

import { requireWeak } from 'utils/webpack';

const server = utils.isomorphy.isServerSide() ?
  requireWeak('topcoder-react-utils/dist/server') : null;

module.exports = {
  client,
  Avatar,
  Button,
  Link,
  NavLink,
  ScalableRect,
  server,
  ...utils,
};
