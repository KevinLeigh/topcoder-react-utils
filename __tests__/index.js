import * as exports from '../src';

test('Exports expected stuff', () => {
  expect(exports).toMatchSnapshot();
});
