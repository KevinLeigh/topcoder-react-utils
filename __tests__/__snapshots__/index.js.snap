// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Export at client side 1`] = `
Object {
  "Avatar": [Function],
  "Button": [Function],
  "JU": null,
  "Link": [Function],
  "NavLink": [Function],
  "ScalableRect": [Function],
  "client": [Function],
  "config": undefined,
  "isomorphy": Object {
    "buildTimestamp": [Function],
    "isClientSide": [Function],
    "isDevBuild": [Function],
    "isProdBuild": [Function],
    "isServerSide": [Function],
  },
  "server": null,
  "webpack": Object {
    "default": undefined,
    "requireWeak": [Function],
  },
}
`;

exports[`Export at server side 1`] = `
Object {
  "Avatar": [Function],
  "Button": [Function],
  "JU": Object {
    "findInDomByClass": [Function],
    "findInDomManyByClass": [Function],
    "render": [Function],
    "renderDom": [Function],
    "shallowRender": [Function],
    "shallowSnapshot": [Function],
    "simulate": Object {
      "abort": [Function],
      "animationEnd": [Function],
      "animationIteration": [Function],
      "animationStart": [Function],
      "beforeInput": [Function],
      "blur": [Function],
      "canPlay": [Function],
      "canPlayThrough": [Function],
      "cancel": [Function],
      "change": [Function],
      "click": [Function],
      "close": [Function],
      "compositionEnd": [Function],
      "compositionStart": [Function],
      "compositionUpdate": [Function],
      "contextMenu": [Function],
      "copy": [Function],
      "cut": [Function],
      "doubleClick": [Function],
      "drag": [Function],
      "dragEnd": [Function],
      "dragEnter": [Function],
      "dragExit": [Function],
      "dragLeave": [Function],
      "dragOver": [Function],
      "dragStart": [Function],
      "drop": [Function],
      "durationChange": [Function],
      "emptied": [Function],
      "encrypted": [Function],
      "ended": [Function],
      "error": [Function],
      "focus": [Function],
      "input": [Function],
      "invalid": [Function],
      "keyDown": [Function],
      "keyPress": [Function],
      "keyUp": [Function],
      "load": [Function],
      "loadStart": [Function],
      "loadedData": [Function],
      "loadedMetadata": [Function],
      "mouseDown": [Function],
      "mouseEnter": [Function],
      "mouseLeave": [Function],
      "mouseMove": [Function],
      "mouseOut": [Function],
      "mouseOver": [Function],
      "mouseUp": [Function],
      "paste": [Function],
      "pause": [Function],
      "play": [Function],
      "playing": [Function],
      "progress": [Function],
      "rateChange": [Function],
      "reset": [Function],
      "scroll": [Function],
      "seeked": [Function],
      "seeking": [Function],
      "select": [Function],
      "stalled": [Function],
      "submit": [Function],
      "suspend": [Function],
      "timeUpdate": [Function],
      "toggle": [Function],
      "touchCancel": [Function],
      "touchEnd": [Function],
      "touchMove": [Function],
      "touchStart": [Function],
      "transitionEnd": [Function],
      "volumeChange": [Function],
      "waiting": [Function],
      "wheel": [Function],
    },
    "snapshot": [Function],
  },
  "Link": [Function],
  "NavLink": [Function],
  "ScalableRect": [Function],
  "client": [Function],
  "config": Config {
    "SECRET": Object {
      "dummySecretKey": "Dummy Secret Value",
    },
    "dummyKey": "Dummy Value",
  },
  "isomorphy": Object {
    "buildTimestamp": [Function],
    "isClientSide": [Function],
    "isDevBuild": [Function],
    "isProdBuild": [Function],
    "isServerSide": [Function],
  },
  "server": [Function],
  "webpack": Object {
    "default": undefined,
    "requireWeak": [Function],
  },
}
`;
